code/sonalyze/db/cache.go

```diff
@@ -0,0 +1,223 @@
+// LogFile content caching logic.
+//
+// A LogFile can choose to cache its content (in whatever form it wants).
+//
+// - To cache some data with a recorded size, call lf.cacheWriteLocked()
+// - To obtain the data (if still there), call lf.cacheReadLocked()
+// - To purge the data, call lf.cachePurgeLocked()
+// - To check whether there are data, call lf.isCachedLocked()
+//
+// Behind the scenes, there is a cache budget, tracking the amount of data available before the
+// cache is full.  Should a write to the cache bring the cache budget below zero, the cache will
+// schedule a background purge.  The purge will pick an item (a LogFile with cached data) to purge
+// from all the files that have cached data, and then purge it by clearing its cache.  It will
+// repeat this until the cache budget is nonnegative.
+//
+// Note that since the purging is asynchronous, it is never possible to know the cached state of a
+// file unless the lock is held.
+//
+// The LogFile.whateverLocked() methods must be called with the file's lock held.
+//
+// Additionally, there is a data structure for managing background purging that has a separate lock,
+// the purgeLock.  The fields of the LogFile that participate in purging are covered *also* by the
+// purgeLock.  Asynchronous purging holds only the purgeLock, all other operations hold first the
+// file lock and then the purgeLock.
```

Comment:
Clarify this: async purging holds the purgelock while picking a file to purge, but will not hold it when it calls on the file to purge itself.

Reply:
In general, *somewhere* we need to document the lock hierarchy.  This was not done for the sonalyzed integration, since more locks were coming for caching.